### ISSUE TITLE
twisted.web server sends responses without explicit Content-Type; potential XSS

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4900]
|<img alt="ivank's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| ivank reported|
|-|-|
|Trac ID|trac#4900|
|Type|defect|
|Created|2011-02-18 03:41:03Z|
|Branch|https://github.com/twisted/twisted/tree/no-utf7-4900|

`twisted.web` serves error pages, directory indexes, redirects, etc without an explicit Content-Type.  If the user can convince the server to include something like `+ADw-script+AD4-alert(+ACc-xss+ACc-)+ADw-+AC8-script+AD4-`, you may have an XSS hole because some browsers will guess that the page encoding is UTF-7.  See http://nedbatchelder.com/blog/200704/xss_with_utf7.html

I think the best practice is to include both a Content-Type header and a Content-Type meta tag (in case the header is somehow stripped, or the page is saved to disk).

Note that `twisted.web.server.Request.process` does `self.setHeader('content-type', "text/html")`.  Changing this would not be backwards-compatible, but most of the other things can be fixed.

<details><summary>Searchable metadata</summary>

```
trac-id__4900 4900
type__defect defect
reporter__ivank ivank
priority__normal normal
milestone__ 
branch__branches_no_utf7_4900 branches/no-utf7-4900
branch_author__glyph glyph
status__closed closed
resolution__fixed fixed
component__web web
keywords__security security
time__1298000463000000 1298000463000000
changetime__1300829026000000 1300829026000000
version__None None
owner__glyph glyph
cc__jknight cc__ivank
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)<a name="note_2"></a>|@jyknight commented|
|-|-|

Note that it's a XSS bug on IE, version 7 and earlier only. Other browsers don't detect UTF-7.

But yes, I agree, anyways: all the builtin content generation should explicitly specify utf-8.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_4"></a>|@glyph commented|
|-|-|

(In [31380]) Branching to 'no-utf7-4900'

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set owner to @glyph<br>@glyph set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_6"></a>|@glyph removed owner<br>@glyph set status to `new`|
|-|-|

[Build Results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/no-utf7-4900)
[Diff](http://twistedmatrix.com:12435/4900)

**Comment by twisted-trac** [comment]
|[<img alt="mwhudson's avatar" src="https://avatars.githubusercontent.com/u/672022?s=50" width="50" height="50">](https://github.com/mwhudson)<a name="note_7"></a>|@mwhudson set owner to @glyph|
|-|-|

This looks fine, please land.

If you _want_ to add a docstring to redirectTo, I'm not going to stop you :)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_8"></a>|@glyph set status to `closed`|
|-|-|

(In [31388]) Merge no-utf7-4900

Author: glyph

Reviewer: mwh

Fixes: [#4900](https://github.com/twisted/twisted/issues/4900)

This change makes error pages and directory listings explicitly set their
encodings in the content-type HTTP header, which should prevent certain
misguided browsers from potentially mis-detecting the encoding and allowing for
inclusion of invalid / insecure content from users.
